tensorflow/core/kernels/non_max_suppression_op.cc

```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
   std::copy_n(selected.begin(), selected.size(), output_indices_data.data());
 }
 
+void BatchedNonMaxSuppressionOp(OpKernelContext* context,
+                                const Tensor& inp_boxes,
+                                const Tensor& inp_scores,
+                                int num_boxes,
+                                const int max_size_per_class,
+                                const int total_size_per_batch,
+                                const float score_threshold,
+                                const float iou_threshold,
+                                bool pad_per_class = false) {
+
+  int q = inp_boxes.dim_size(2);
+  int num_classes = inp_scores.dim_size(2);
+  //unpack along batch dimension
+  const int num_batches = inp_boxes.dim_size(0);
+
+  // Default clip window of [0, 0, 1, 1] if none specified
+  std::vector<float> clip_window{0, 0, 1, 1};
+  
+  // [num_batches, per_batch_size, 4]
+  std::vector<std::vector<float>> nmsed_boxes(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_scores(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_classes(num_batches);
+  // [num_batches]
+  std::vector<int> final_valid_detections;
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<int>> selected_indices(num_batches);
+
+  int per_batch_size = total_size_per_batch;
+
+  //perform non_max_suppression operation for each batch independently
+  for (int batch = 0; batch < num_batches; ++batch) {
+    // dims of per_batch_boxes [num_boxes, q, 4]
+    Tensor per_batch_boxes = inp_boxes.Slice(batch, batch+1);
+    // dims of per_batch_scores [num_boxes, num_classes]
+    Tensor per_batch_scores = inp_scores.Slice(batch, batch+1);
+
+    struct ResultCandidate {
+      int box_index;
+      float score;
+      int class_idx;
+      float box_coord[4];
+    };
+
+    auto rc_cmp = [](const ResultCandidate rc_i, const ResultCandidate rc_j) {
+        return rc_i.score < rc_j.score;
+    };
+    std::priority_queue<ResultCandidate, std::vector<ResultCandidate>, 
+      decltype(rc_cmp)> result_candidate_pq(rc_cmp);
+
+    float * scoresData = per_batch_scores.unaligned_flat<float>().data();
+    float * boxesData = per_batch_boxes.unaligned_flat<float>().data();
+
+    //Iterate through all classes
+    for (int class_idx = 0; class_idx < num_classes; ++class_idx) {
+      std::vector<float> scores_data;
+      std::vector<float> boxes_data_vec;
+
+      for (int box = 0; box < num_boxes; ++box) {
+        //Get the scores per class 
+        //scores_data dim is [num_boxes].
+        scores_data.push_back(scoresData[box * num_classes + class_idx]);
```

Comment:
can we avoid the copy. especially since you copy back boxes to a tensor afterword

Reply:
I am not sure how to do this since I need to slice along the column of the matrix. Do you have any suggestions on how it can be done?